web/src/lib/components/memory-page/memory-viewer.svelte

```diff
@@ -0,0 +1,354 @@
+	import Play from 'svelte-material-icons/Play.svelte';
+	import Pause from 'svelte-material-icons/Pause.svelte';
+	import ChevronDown from 'svelte-material-icons/ChevronDown.svelte';
+	import ChevronUp from 'svelte-material-icons/ChevronUp.svelte';
+	import ChevronLeft from 'svelte-material-icons/ChevronLeft.svelte';
+	import ChevronRight from 'svelte-material-icons/ChevronRight.svelte';
+	import { AppRoute } from '$lib/constants';
+	import { page } from '$app/stores';
+	import noThumbnailUrl from '$lib/assets/no-thumbnail.png';
+	import GalleryViewer from '$lib/components/shared-components/gallery-viewer/gallery-viewer.svelte';
+	import CircleIconButton from '$lib/components/elements/buttons/circle-icon-button.svelte';
+	import IntersectionObserver from '$lib/components/asset-viewer/intersection-observer.svelte';
+	import { fade } from 'svelte/transition';
+
+	let currentIndex = 0;
+	let currentMemory: MemoryLaneResponseDto;
+	let nextMemory: MemoryLaneResponseDto;
+	let lastMemory: MemoryLaneResponseDto;
+
+	let lastIndex = 0;
+	let nextIndex = 0;
+	$: showNextMemory = nextIndex <= $memoryStore?.length - 1;
+	$: showPreviousMemory = currentIndex != 0;
+
+	let memoryGallery: HTMLElement;
+	let memoryWrapper: HTMLElement;
+	let galleryInView = false;
+
+	onMount(async () => {
+		if (!$memoryStore) {
+			const timezone = DateTime.local().zoneName;
+			const { data } = await api.assetApi.getMemoryLane({ timezone });
+			$memoryStore = data;
+		}
+
+		const queryIndex = $page.url.searchParams.get('index');
+		if (queryIndex != null) {
+			currentIndex = parseInt(queryIndex);
+			if (isNaN(currentIndex) || currentIndex > $memoryStore.length - 1) {
+				currentIndex = 0;
+			}
+		}
+
+		currentMemory = $memoryStore[currentIndex];
+
+		nextIndex = currentIndex + 1;
+		nextMemory = $memoryStore[nextIndex];
+
+		if (currentIndex > 0) {
+			lastMemory = $memoryStore[lastIndex];
+		}
+	});
+
+	const toNextMemory = (): boolean => {
+		if (showNextMemory) {
+			resetAutoPlay();
+
+			currentIndex += 1;
+			nextIndex = currentIndex + 1;
+			lastIndex = currentIndex - 1;
+
+			currentMemory = $memoryStore[currentIndex];
+			nextMemory = $memoryStore[nextIndex];
+			lastMemory = $memoryStore[lastIndex];
+
+			return true;
+		}
+
+		return false;
+	};
+
+	const toPreviousMemory = () => {
+		if (showPreviousMemory) {
+			resetAutoPlay();
+
+			currentIndex -= 1;
+			nextIndex = currentIndex + 1;
+			lastIndex = currentIndex - 1;
+
+			currentMemory = $memoryStore[currentIndex];
+			nextMemory = $memoryStore[nextIndex];
+			lastMemory = $memoryStore[lastIndex];
+		}
+	};
+
+	let autoPlayInterval: NodeJS.Timeout;
+	let autoPlay = false;
+	let autoPlaySpeed = 5000;
+	let autoPlayProgress = 0;
+	let autoPlayIndex = 0;
+	let canPlayNext = true;
+
+	const toggleAutoPlay = () => {
+		autoPlay = !autoPlay;
+		if (autoPlay) {
+			autoPlayInterval = setInterval(() => {
+				if (!canPlayNext) return;
+
+				window.requestAnimationFrame(() => {
```

Comment:
What's the purpose of an animation frame here?

Reply:
I believe it is optimized for animation on screen

You can search around the internet for more information. I used this technique from implementing the virtual viewport

You can read the "Minimize changes" section in the article below

https://medium.com/google-design/google-photos-45b714dfbed1

Reply:
That's a nice article, and even talks a bit about low resolution thumbnails which is relevant to some other work going on in #2649.

Though I still don't quite get the actual use of animation frame here. An animation frame is being request to increment an integer? I would imagine that won't do much if anything at all, and instead more logic would need to be moved into the animation frame. I could be mistaken though.

Reply:
The biggest thing it will do is not update the progress bar when the page is in the background.

Reply:
Well, it's still not great as suddenly all the animation frames are run all at once when the page is moved to the foreground again.

Reply:
The behaviour also becomes really strange if the animation frame doesn't run in reasonable time, all those animation frames will run all at once as mentioned before and I'm not sure how svelte would deal with that. Would it quickly do all steps from 0-100? Seems like it needs some thinking.

Reply:
Assuming the intended behaviour is for the animation to pause when the page is moved to the foreground, this may make more sense:

```ts
function draw() {
  setTimeout(() => requestAnimationFrame(() => {}), autoPlaySpeed / 100);
}
draw();
```

Alternatively, it could be better to use requestAnimationFrame exlusively. It would be very efficient.

```ts
let previousTransition;
function draw(now) {
  previousTransition ??= now;

  const elapsed = previousTransition - now;

  autoPlayProgress = elapsed / autoPlaySpeed;
  previousTransition = now;

  // etc ...

  requestAnimationFrame(draw);
}
requestAnimationFrame(draw);
```

Reply:
It would probably be worth using [cancelAnimationFrame](https://developer.mozilla.org/en-US/docs/Web/API/Window/cancelAnimationFrame) somewhere too to prevent two animation threads from running at the same time if one were to enable/disable auto-play in quick succession.